File: go/url.go
package main

import (
	"bufio"
	"crypto/tls"
	"fmt"
	"io"
	"net"
	"os"
	"slices"
	"strconv"
	str "strings"
)

type Url struct {
	scheme, host, path, port, content_type string
}

func newUrlWithContentType(url string, content_type string) Url {
	scheme, url := split1(url, ":")
	doubleslash_schemes := []string{"http", "https", "file"}
	supported_schemes := []string{"http", "https", "file", "data", "view-source"}

	if !slices.Contains(supported_schemes, scheme) {
		panic(fmt.Errorf("unsupported scheme %s. Supported schemes are %v", scheme, supported_schemes))
	}

	// check for and remove "//" at beginning of http, file, etc
	if slices.Contains(doubleslash_schemes, scheme) {
		if str.HasPrefix(url, "//") {
			url = str.TrimPrefix(url, "//")
		} else {
			panic(fmt.Errorf("malformed uri: %s, expected leading //", url))
		}
	}

	if scheme == "file" {
		if getExt(url) != "html" {
			content_type = "text/ascii"
		}
		return Url{scheme: scheme, host: "", path: url, port: "0", content_type: content_type}
	}

	if scheme == "data" {
		content_type, content := split1(url, ",")
		return Url{scheme: scheme, host: "", path: content, port: "0", content_type: content_type}
	}

	if scheme == "view-source" {
		return newUrlWithContentType(url, "text/ascii")
	}

	host, path := split1(url, "/")
	if !str.HasPrefix(path, "/") {
		path = "/" + path
	}

	// check for port
	host, port := split1(host, ":")

	if len(port) == 0 {
		switch scheme {
		case "http":
			port = "80"
		case "https":
			port = "443"
		default:
			port = "80"
		}
	}

	return Url{scheme: scheme, host: host, path: path, port: port, content_type: content_type}
}

func newUrl(url string) Url {
	return newUrlWithContentType(url, "text/html")
}

func request(url Url) ([]byte, error) {

	if url.scheme == "file" {
		return os.ReadFile(url.path)
	}

	if url.scheme == "data" {
		return []byte(url.path), nil
	}

	address := url.host + ":" + url.port

	var conn net.Conn
	var err error
	if url.scheme == "https" {
		conn, err = tls.Dial("tcp", address, &tls.Config{})
	} else {
		conn, err = net.Dial("tcp", address)
	}

	if err != nil {
		panic(fmt.Errorf("cannot connect to host %s on port %s\n. Error: %v", url.host, url.port, err))
	}

	// send HTTP get request
	req := fmt.Sprintf("GET %s HTTP/1.1\r\n", url.path)
	req += fmt.Sprintf("Host: %s\r\n", url.host)
	req += "Connection: close\r\n"
	req += "User-Agent: archermarx-browser\r\n"
	req += "\r\n"
	fmt.Fprint(conn, req)

	// get status line and parse error code
	response_reader := bufio.NewReader(conn)
	status_line, _ := response_reader.ReadString('\n')
	status_line = str.TrimRight(status_line, "\r\n")
	parts := str.Split(status_line, " ")
	status_code, _ := strconv.Atoi(parts[1])

	if status_code >= 400 {
		panic(fmt.Errorf("bad request\n%s", req))
	}

	response_headers := make(map[string]string)
	for {
		line, _ := response_reader.ReadString('\n')
		if line == "\r\n" {
			break
		}
		line = str.TrimRight(line, "\r\n")
		key, val := split1(line, ":")
		response_headers[str.ToLower(key)] = str.Trim(val, " ")
	}

	if _, ok := response_headers["transfer-encoding"]; ok {
		panic("transfer-encoding not supported")
	}

	if _, ok := response_headers["content-encoding"]; ok {
		panic("content-enconding not supported")
	}

	return io.ReadAll(response_reader)
}

File: py/url.py
import socket
import ssl
import time
from typing import Dict, Tuple
import gzip

response_cache: Dict[str, Tuple[float, float, str]] = {}
class URL:
    
    def url_check(self, cond):
        if not cond:
            self.malformed = True

    def __init__(self, url, redirect_depth = 0):
        # store initial url to use a as a key for cache lookups
        self.url = url
        self.malformed = False

        # prevent infinite redirect chains
        self.redirect_depth = redirect_depth

        # set default content type
        self.content_type = "text/html"

        # determine scheme - only http, https, file, data, and view-source supported for now
        # if no scheme provided, file is assumed

        # check for view-source scheme
        if url.startswith("view-source:"):
            self.content_type = "text/plain;charset=utf8"
            url = url[12:]

        if "://" in url:
            self.scheme, url = url.split("://", 1)
            self.url_check(self.scheme in ["http", "https", "file"])
        elif ":" in url:
            self.scheme, url = url.split(":", 1)
            self.url_check(self.scheme == "data")
        else:   
            self.scheme = "file"

        # handle file scheme
        if self.scheme == "file":
            self.path = url
            if "." in self.path:
                _, ext = self.path.split(".", 1)
                if ext != "html":
                    self.content_type = "text/plain;charset=utf8"
            return
        
        if self.scheme == "data":
            self.url_check("," in url)
            self.content_type, self.data = url.split(",", 1)
            if self.content_type == "":
                self.content_type = "text/plain;charset=US-ASCII"

        if self.malformed:
            self.url = "about:blank"
            return;

        # separate host from path
        if "/" not in url:
            url = url + "/"

        self.host, url = url.split("/", 1)
        self.path = "/" + url 

        # determine which port we're using
        if ":" in self.host:
            # custom port present
            self.host, port = url.split(":", 1)
            self.port = int(port)
        elif self.scheme == "http":
            self.port = 80
        elif self.scheme == "https":
            self.port = 443

        self.socket = None

    def request(self):
        if self.malformed:
            return "", None, None

        if self.scheme == "data":
            return self.data, None, None

        if self.scheme == "file":
            with open(self.path) as f:
                return f.read(), None, None

        current_time = time.time()
            
        if self.url in response_cache:
            # use cached response if page not too old
            content = response_cache[self.url]
            timestamp, max_age, content = response_cache[self.url]
            age = current_time - timestamp
            if (max_age < 0) or (age < max_age):
                return content, True, max_age
      
        #  Create socket
        self.socket = socket.socket(
            family = socket.AF_INET,
            type = socket.SOCK_STREAM,
            proto = socket.IPPROTO_TCP,
        )

        try:
            self.socket.connect((self.host, self.port))
        except:
            return "Invalid URL", None, None
        
        # wrap socket with SSL encryption if using https
        if self.scheme == "https":
            ctx = ssl.create_default_context()
            self.socket = ctx.wrap_socket(self.socket, server_hostname = self.host)

        # send HTTP GET request
        request = f"GET {self.path} HTTP/1.1\r\n"
        request += f"Host: {self.host}\r\n"
        request += "User-Agent: archermarx\r\n"
        request += "Connection: close\r\n"
        request += "Accept-Encoding: gzip\r\n"
        request += "\r\n"
        self.socket.send(request.encode("utf8"))            
        
        response = self.socket.makefile("rb", newline = "\r\n")

        # Parse the response
        statusline = response.readline().decode("utf8")
        version, status, explanation = statusline.split(" ", 2)
        status = int(status)

        response_headers = {}
        while True:
            line = response.readline().decode("utf8")
            if line == "\r\n": 
                break
            header, value = line.split(":", 1)
            response_headers[header.casefold()] = value.strip()

        max_age = -1.0
        cache = True

        if "cache-control" in response_headers:
            options = response_headers["cache-control"].split(",")
            for opt in options:
                opt = opt.strip()
                if opt == "no-store":
                    cache = False
                elif opt.startswith("max-age"):
                    _, val = opt.split("=", 1)
                    max_age = float(val)
                else:
                    # don't cache if other options provided
                    cache = False
            
        # don't bother caching pages with very short max ages
        if max_age < 1.0:
            cache = False

        # check for redirect
        if 300 <= status < 400 and self.redirect_depth < 10:
            location = response_headers["location"]
            if location.startswith("/"):
                location = self.scheme + "://" + self.host + location

            print(f"Redirecting to {location} (depth = {self.redirect_depth})")
            redirect = URL(location, self.redirect_depth + 1)
            # todo: cache redirects without caching whole page
            return redirect.request()

        self.content_type = response_headers["content-type"]

        chunked = False
        compressed = False

        if "transfer-encoding" in response_headers:
            transfer_encoding_opts = [s.strip() for s in response_headers["transfer-encoding"].split(",")]
            chunked = "chunked" in transfer_encoding_opts
            compressed = "gzip" in transfer_encoding_opts

        if "content-encoding" in response_headers:
            content_encoding_opts = [s.strip() for s in response_headers["content-encoding"].split(",")]
            compressed = "gzip" in content_encoding_opts

        def read_chunk(f):
            chunk = b""
            chunk_size = f.readline()
            chunk_size = int(chunk_size[:-2], 16)
            if chunk_size > 0:
                chunk += f.read(chunk_size)
                f.readline()
            return chunk

        # read content in chunks if requested
        if chunked:
            content = b""
            while True:
                chunk = read_chunk(response)
                if not chunk:
                    break
                content += chunk

        else:
            bytes_to_read = int(response_headers["content-length"])
            content = response.read(bytes_to_read)

        # decompress content if needed
        if compressed:
            content = gzip.decompress(content)

        content = content.decode("utf-8")

        # cache content
        if cache and (status == 200 or status == 404):
            response_cache[self.url] = (current_time, max_age, content)

        self.socket.close()

        return content, cache, max_age
